simpleclient/src/main/java/io/prometheus/client/Gauge.java

```diff
@@ -234,9 +231,10 @@ public double setToTime(Runnable timeable){
      * Get the value of the gauge.
      */
     public double get() {
-      synchronized(this) {
-        return value.sum();
-      }
+      // On concurrent `get` and `set`, it is acceptable to `get` an outdated `value`.
+      // On concurrent `get` and `inc / dec`, it is acceptable to `get` an outdated `value`.
+      // On concurrent `get` and `set` and `inc / dec`, it is possible to `get` an invalid `value`.
```

Comment:
I believe this tradeoff should be acceptable, as mixed `set` and `inc / dec` use cases should be rare.

Reply:
> On concurrent `get` and `set` and `inc / dec`, it is possible to `get` an invalid `value`.

This is not acceptable. You should get some value that some serialisation of the set/inc/get could produce.

Reply:
The invalid value is when `sum` reads the new `cells` and old `base`.

Let me try to think about if it's possible to block  `sum` on concurrent `set` without affecting anything else.

Reply:
I think I managed to fix it with busy waiting / retry loop based on the idea of optimistic concurrency / [Software Transactional Memory](https://en.wikipedia.org/wiki/Software_transactional_memory).